File: lib/src/main/java/fr/o80/sample/lib/core/presenter/Presenter.java
package fr.o80.sample.lib.core.presenter;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * @author Olivier Perez
 */
public class Presenter<T extends PresenterView> {

    private T view;

    private CompositeDisposable disposables = new CompositeDisposable();

    public void attach(T view) {
        this.view = view;
    }

    public void dettach() {
        disposables.dispose();
        view = null;
    }

    public T getView() {
        if (view == null) {
            throw new RuntimeException("Trying to access view after it has been dettached.");
        }
        return view;
    }

    public T getViewUnsafe() {
        return view;
    }

    protected void addDisposable(Disposable disposable) {
        disposables.add(disposable);
    }
}

File: lib/src/main/java/fr/o80/sample/lib/core/presenter/Presenter.kt
package fr.o80.sample.lib.core.presenter

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

/**
 * @author Olivier Perez
 */
open class Presenter<out T : PresenterView> {

    private var viewUnsafe: PresenterView? = null

    private val disposables = CompositeDisposable()

    fun attach(view: PresenterView?) {
        this.viewUnsafe = view
    }

    fun dettach() {
        disposables.clear()
        viewUnsafe = null
    }

    @Suppress("UNCHECKED_CAST")
    val view: T
        get() {
            when (viewUnsafe) {
                null -> throw RuntimeException("Trying to access view after it has been dettached.")
                else -> return viewUnsafe as T
            }
        }

    protected fun addDisposable(disposable: Disposable) {
        disposables.add(disposable)
    }
}
